File: backend/modules/network/protocol.py
"""Peer-wire serialization, signing, and the loop/replay guard.

Canonicalization must be **byte-stable across nodes** (sorted keys, fixed
separators) so a signature produced on one machine verifies on another. The
`SeenGuard` deduplicates `msg_id`s and the envelope `ttl` bounds relay hops — the
two together stop loops and replays on the fabric.
"""

from __future__ import annotations

import json
from collections import OrderedDict

from backend.modules.network import identity
from backend.modules.network.models import PeerEnvelope

# Message-type constants (the `type` field of a PeerEnvelope).
HELLO = "hello"
HELLO_ACK = "hello_ack"
AUTH = "auth"
AUTH_RESULT = "auth_result"
PRESENCE = "presence"
PING = "ping"
PONG = "pong"
AGENT_REQUEST = "agent_request"
AGENT_STREAM = "agent_stream"
AGENT_RESULT = "agent_result"
AGENT_CANCEL = "agent_cancel"
COLLAB_JOIN = "collab_join"
COLLAB_LEAVE = "collab_leave"
COLLAB_OP = "collab_op"
ERROR = "error"


def canonical_bytes(env: PeerEnvelope) -> bytes:
    """The bytes an envelope signature covers: the authenticated fields only,
    encoded with sorted keys and compact separators for cross-node byte stability.

    `dst` and `ttl` are routing headers a relay may legitimately set/rewrite while
    forwarding, so they are excluded — the recipient still authenticates `src`,
    `type`, `msg_id`, `re`, `ts`, and `data`."""
    payload = env.model_dump(exclude={"sig", "dst", "ttl"})
    return json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")


def sign_envelope(
    env: PeerEnvelope, signer: identity.Identity | None = None
) -> PeerEnvelope:
    """Return a copy of `env` signed by `signer` (this node by default; pass an
    explicit identity so a hub signs as itself regardless of process-global state).
    Also stamps `src` if unset."""
    me = signer or identity.load_identity()
    if not env.src:
        env = env.model_copy(update={"src": me.node_id})
    env.sig = me.sign(canonical_bytes(env))
    return env


def verify_envelope(env: PeerEnvelope, public_key_b64: str) -> bool:
    """Whether `env.sig` is a valid signature over the envelope by `public_key_b64`.
    The caller is responsible for checking the key's fingerprint matches `env.src`."""
    if not env.sig:
        return False
    return identity.verify(public_key_b64, canonical_bytes(env), env.sig)


def encode(env: PeerEnvelope) -> str:
    return env.model_dump_json()


def decode(raw: str) -> PeerEnvelope:
    return PeerEnvelope.model_validate_json(raw)


class SeenGuard:
    """Bounded LRU of recently-seen `msg_id`s. `check(msg_id)` returns True the first
    time an id is seen and False on a repeat, so a replayed/looped envelope is
    dropped."""

    def __init__(self, capacity: int = 4096) -> None:
        self._seen: OrderedDict[str, None] = OrderedDict()
        self._capacity = capacity

    def check(self, msg_id: str) -> bool:
        if msg_id in self._seen:
            self._seen.move_to_end(msg_id)
            return False
        self._seen[msg_id] = None
        if len(self._seen) > self._capacity:
            self._seen.popitem(last=False)
        return True

File: apps/mobile-android/app/src/main/kotlin/com/horrible/dashboard/network/Protocol.kt
package com.horrible.dashboard.network

import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import org.json.JSONObject

/**
 * Peer-wire serialization + signing. Mirrors `backend/modules/network/protocol.py`.
 *
 * The signature covers the **authenticated fields only** — `sig`, `dst` and `ttl`
 * are excluded because a relay may legitimately rewrite the latter two in flight.
 * Canonicalization runs over the *raw wire JSON* rather than the parsed
 * [PeerEnvelope]: Moshi widens every JSON number to `Double`, so round-tripping an
 * inbound envelope through Kotlin objects would turn the desktop's `"v":1` into
 * `1.0` and fail every verification. See [CanonicalJson].
 */
object Protocol {
    private val moshi = Moshi.Builder()
        .addLast(KotlinJsonAdapterFactory())
        .build()
    private val envelopeAdapter = moshi.adapter(PeerEnvelope::class.java)

    // Message types — keep in sync with backend/modules/network/protocol.py.
    const val HELLO = "hello"
    const val HELLO_ACK = "hello_ack"
    const val AUTH = "auth"
    const val AUTH_RESULT = "auth_result"
    const val PRESENCE = "presence"
    const val PING = "ping"
    const val PONG = "pong"
    const val AGENT_REQUEST = "agent_request"
    const val AGENT_STREAM = "agent_stream"
    const val AGENT_RESULT = "agent_result"
    const val AGENT_CANCEL = "agent_cancel"
    const val PEER_CHAT = "peer_chat"
    const val REMOTE_COMMAND = "remote_command"
    const val VIEW_REQUEST = "view_request"
    const val VIEW_ACCEPT = "view_accept"
    const val VIEW_FRAME = "view_frame"
    const val ERROR = "error"

    fun encode(env: PeerEnvelope): String = envelopeAdapter.toJson(env)

    fun decode(raw: String): PeerEnvelope? = envelopeAdapter.fromJson(raw)

    /**
     * The bytes a signature covers, for an envelope in its raw wire form.
     *
     * Moshi omits null fields, but Python's `model_dump` always emits `re` (as
     * `null`) — so it is added back when absent. Everything else follows from
     * [CanonicalJson] reproducing `json.dumps(sort_keys=True, separators=(",",":"))`.
     */
    fun canonicalBytes(rawJson: String): ByteArray {
        val obj = JSONObject(rawJson)
        obj.remove("sig")
        obj.remove("dst")
        obj.remove("ttl")
        if (!obj.has("re")) obj.put("re", JSONObject.NULL)
        return CanonicalJson.dumps(obj).toByteArray(Charsets.UTF_8)
    }

    fun signEnvelope(env: PeerEnvelope, identity: Identity): PeerEnvelope {
        val unsigned = if (env.sig == null) env else env.copy(sig = null)
        val sig = identity.sign(canonicalBytes(encode(unsigned)))
        return env.copy(sig = sig)
    }

    /** Whether `rawJson`'s `sig` is a valid signature over it by `publicKeyB64`. */
    fun verifyEnvelope(rawJson: String, publicKeyB64: String): Boolean {
        return try {
            val sig = JSONObject(rawJson).optString("sig", "")
            if (sig.isEmpty()) false
            else Identity.verify(publicKeyB64, canonicalBytes(rawJson), sig)
        } catch (e: Exception) {
            false
        }
    }
}
